File: automcp/adapters/mcp_agent.py
from typing import Any, Callable, Type
from pydantic import BaseModel
import asyncio
import inspect


def create_mcp_agent_adapter(
    agent_instance: Any,
    llm: Any,
    app: Any,
    app_initialize_fn: Callable,
    name: str,
    description: str,
    input_schema: Type[BaseModel],
) -> Callable:
    """
    Convert a MCP agent instance to an MCP tool with proper isolation from MCP's task management.
    """
    # Define the wrapper function
    async def run_agent(**kwargs):
        # Create input object from schema
        input_data = input_schema(**kwargs)
        
        # Create a future that we can use to communicate between tasks
        result_future = asyncio.Future()
        
        async def isolated_agent_task():
            """Run the agent in an isolated context"""
            try:
                # Initialize the app
                await app_initialize_fn(app)
                
                # Attach LLM to the agent
                llm_instance = await agent_instance.attach_llm(llm)
                
                # Execute the main operation
                response = await llm_instance.generate_str(input_data.query)
                
                # Set the result if the future is not done yet
                if not result_future.done():
                    result_future.set_result(response)
            except Exception as e:
                # Set exception if the future is not done yet
                if not result_future.done():
                    result_future.set_exception(e)
        
        # Run the agent task in a new task
        task = asyncio.create_task(isolated_agent_task())
        
        try:
            # Wait for the result or an exception
            return await result_future
        except asyncio.CancelledError:
            # If our wrapper gets cancelled, cancel the agent task too
            task.cancel()
            raise
        except Exception as e:
            # Log any other exceptions but don't re-raise
            import traceback
            traceback.print_exc()
            return {"status": "error", "message": str(e)}
    
    # Create a function with the proper signature
    params = []
    for field_name, field_info in input_schema.model_fields.items():
        param = inspect.Parameter(
            name=field_name,
            kind=inspect.Parameter.POSITIONAL_OR_KEYWORD,
            annotation=field_info.annotation
        )
        params.append(param)
    
    run_agent.__signature__ = inspect.Signature(parameters=params)
    run_agent.__name__ = name
    run_agent.__doc__ = description
    
    return run_agent

File: src/adapters/mcpAgent.ts
import { ToolResult } from './base.js';
import { ensureSerializable, extractParamsFromArgs, parseInputWithSchema, type SchemaLike } from './utils.js';

interface BaseModel {
  [key: string]: any;
}

interface AgentInstance {
  attach_llm(llm: any): Promise<any>;
}

interface LLMInstance {
  generate_str(query: string): Promise<string>;
}

type AppInitializeFn = (app: any) => Promise<void>;

/**
 * Convert a MCP agent instance to an MCP tool with proper isolation from MCP's task management.
 */
export function createMcpAgentAdapter(
  agentInstance: AgentInstance,
  llm: any,
  app: any,
  appInitializeFn: AppInitializeFn,
  name: string,
  description: string,
  inputSchema: SchemaLike
): (...args: any[]) => Promise<ToolResult> {
  // Define the wrapper function
  const runAgent = async (...args: any[]): Promise<ToolResult> => {
    const kwargs = extractParamsFromArgs(args, inputSchema);
    const inputData = parseInputWithSchema(inputSchema, kwargs);
    const queryInput = inputData && typeof inputData === 'object'
      ? (inputData as Record<string, any>).query ?? inputData
      : inputData;

    // Create a promise that we can use to communicate between tasks
    let resolveResult: (value: any) => void;
    let rejectResult: (reason: any) => void;
    const resultPromise = new Promise<any>((resolve, reject) => {
      resolveResult = resolve;
      rejectResult = reject;
    });
    
    const isolatedAgentTask = async (): Promise<void> => {
      try {
        // Initialize the app
        await appInitializeFn(app);
        
        // Attach LLM to the agent
        const llmInstance: LLMInstance = await agentInstance.attach_llm(llm);

        // Execute the main operation
        const response = await llmInstance.generate_str(
          typeof queryInput === 'string' ? queryInput : JSON.stringify(queryInput)
        );

        // Set the result
        resolveResult(response);
      } catch (error) {
        // Set exception
        rejectResult(error);
      }
    };
    
    // Start the agent task
    const taskPromise = isolatedAgentTask();
    
    try {
      // Wait for the result or an exception
      const result = await Promise.race([
        resultPromise,
        taskPromise.then(() => resultPromise) // Ensure we wait for the task to complete
      ]);
      const output = ensureSerializable(result);
      if (output && typeof output === 'object') {
        return {
          content: [{ type: 'text', text: JSON.stringify(output, null, 2) }],
          structuredContent: output
        };
      }
      return { content: [{ type: 'text', text: String(result) }] };
    } catch (error) {
      // Log any exceptions but don't re-raise unless it's a cancellation
      if (error instanceof Error) {
        console.error('Agent execution error:', error.stack || error.message);
        return {
          content: [{ type: 'text', text: error.message }],
          isError: true
        };
      }
      throw error;
    }
  };
  
  // Set function metadata
  Object.defineProperty(runAgent, 'name', { value: name });
  Object.defineProperty(runAgent, 'description', { value: description });
  
  return runAgent;
}

/**
 * Helper function to create a schema-based function signature
 */
export function createSchemaFunction<T extends BaseModel>(
  schema: SchemaLike,
  implementation: (input: T) => Promise<any>
): (data: any) => Promise<any> {
  return async (data: any) => {
    const params = typeof data === 'object' && data !== null ? data : { value: data };
    const input = parseInputWithSchema(schema, params) as T;
    return await implementation(input);
  };
}